mesa/model.py

```diff
@@ -65,27 +66,39 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self.running = True
         self.schedule = None
         self.current_id = 0
-        self._agents: defaultdict[type, dict] = defaultdict(dict)
+        self.agents_: defaultdict[type, dict] = defaultdict(dict)
 
         # Warning flags for current experimental features. These make sure a warning is only printed once per model.
         self.agentset_experimental_warning_given = False
 
     @property
     def agents(self) -> AgentSet:
         """Provides an AgentSet of all agents in the model, combining agents from all types."""
-        all_agents = itertools.chain(
-            *(agents_by_type.keys() for agents_by_type in self._agents.values())
-        )
-        return AgentSet(all_agents, self)
+
+        if hasattr(self, "_agents"):
+            return self._agents
+        else:
+
+            all_agents = itertools.chain(
+                *(agents_by_type.keys() for agents_by_type in self.agents_.values())
+            )
+            return AgentSet(all_agents, self)
+
+    @agents.setter
+    def agents(self, agents: Any) -> None:
+        warnings.warn("You are trying to set model.agents. In a next release, this attribute is used "
+                      "by MESA itself so you cannot use it directly anymore.", UserWarning)
```

Comment:
maybe add "Please adjust your code to use a different attribute name for custom agent storage."